Allow User-supplied typemaps to take precedence

bay-max1 opened the pull request:
According to

https://github.com/Perl-Toolchain-Gang/extutils-parsexs/blob/master/README#L64-L68:

>       typemap
>            Indicates that a user-supplied typemap should take precedence
>            over the default typemaps. A single typemap may be specified as
>            a string, or multiple typemaps can be specified in an array
>            reference, with the last typemap having the highest precedence.

the last element of the array-ref passed to `ExtUtils::ParseXS::Utilities::process_typemaps` should be taking precedence, as `ExtUtils::ParseXS::process_file` just passes it's `$args{typemap}` to `process_typemaps`, shown here:

https://github.com/Perl-Toolchain-Gang/extutils-parsexs/blob/29e5a1ef39513ff5815b3faf8fe8b47d68c72e9a/lib/ExtUtils/ParseXS.pm#L175


But passed typemaps are not actually taking precedence unless it's named 'typemap' and appears in one of the locations listed here:

https://github.com/Perl-Toolchain-Gang/extutils-parsexs/blob/29e5a1ef39513ff5815b3faf8fe8b47d68c72e9a/lib/ExtUtils/ParseXS/Utilities.pm#L102-L111


The problem is, `standard_typemap_locations( \@INC )`, shown here

https://github.com/Perl-Toolchain-Gang/extutils-parsexs/blob/29e5a1ef39513ff5815b3faf8fe8b47d68c72e9a/lib/ExtUtils/ParseXS/Utilities.pm#L275

is `push()`ed to the end of the typemap array, `@tm`, and since the last-most entries take precedence, it's not possible for the supplied typemaps to override entries returned by `standard_typemap_locations()`.

`unshift()`ing instead of `push()`ing would allow supplied typemaps to properly take precedence.

The reason this probably went unnoticed for so long is most likely due to the problem only occurring when overriding a typemap defined in `perl -V:privlib`/ExtUtils/typemap.

For example, when **./typemap-foo** and `perl -V:privlib`**/ExtUtils/typemap** both define a mapping for the same data type, the definition in `perl -V:privlib`**/ExtUtils/typemap**, which is processed after **./typemap-foo**, is the one that is used because the last one wins.

When *./typemap-foo* defines type mappings that are not found in `perl -V:privlib`**/ExtUtils/typemap** (or anywhere else returned by `standard_typemap_locations()`) then they are merged in just fine since they don't exist in `perl -V:privlib`**/ExtUtils/typemap** and thus don't get clobbered.


### Here are examples that demonstrate the problem in ExtUtils::ParseXS 3.35 or earlier. (Run these commands in an empty directory.)

##### Note: `ExtUtils::ParseXS::Utilities::process_typemaps([@ARGV], pwd)` in the examples below are based on this line:
https://github.com/Perl-Toolchain-Gang/extutils-parsexs/blob/29e5a1ef39513ff5815b3faf8fe8b47d68c72e9a/lib/ExtUtils/ParseXS.pm#L175


#### Linux/Unix

`\rm typemap*; echo -e 'TYPEMAP\nwchar_t * T_WCHAR' > typemap; perl -MExtUtils::ParseXS::Utilities -E 'say ExtUtils::ParseXS::Utilities::process_typemaps([@ARGV], pwd)->as_string' typemap | grep -i wchar`
wchar_t T_IV
wchar_t *       T_WCHAR

`\rm typemap*; echo -e 'TYPEMAP\nwchar_t * T_WCHAR' > typemap-other; perl -MExtUtils::ParseXS::Utilities -E 'say ExtUtils::ParseXS::Utilities::process_typemaps([@ARGV], pwd)->as_string' typemap-other | grep -i wchar`
wchar_t *       T_PV
wchar_t T_IV


#### Win32:

`del typemap* & echo TYPEMAP > typemap & echo wchar_t * T_WCHAR >> typemap & perl -MExtUtils::ParseXS::Utilities -E "say ExtUtils::ParseXS::Utilities::process_typemaps([@ARGV], pwd)->as_string" typemap | grep -i wchar`
wchar_t T_IV
wchar_t *       T_WCHAR

`del typemap* & echo TYPEMAP > typemap-other & echo wchar_t * T_WCHAR >> typemap-other & perl -MExtUtils::ParseXS::Utilities -E "say ExtUtils::ParseXS::Utilities::process_typemaps([@ARGV], pwd)->as_string" typemap-other | grep -i wchar`
wchar_t *       T_PV
wchar_t T_IV

##### `wchar_t * T_PV` is defined in `perl -V:privlib`*/ExtUtils/typemap* so in the examples using *typemap-other* instead of *typemap*, it overrides the user-supplied `wchar_t * T_WCHAR` entry.

bay-max1 committed:
Allow User-supplied typemaps to take precedence

According to https://github.com/Perl-Toolchain-Gang/extutils-parsexs/blob/master/README#L64-L68:

"
       typemap
            Indicates that a user-supplied typemap should take precedence
            over the default typemaps. A single typemap may be specified as
            a string, or multiple typemaps can be specified in an array
            reference, with the last typemap having the highest precedence.
"


Therefore, the last argument passed to ExtUtils::ParseXS::Utilities::process_typemaps([@typemaps]) 
should be taking precedence, as ExtUtils::ParseXS::process_file just passes it's $args{typemap} 
to process_typemaps, shown here:

https://github.com/Perl-Toolchain-Gang/extutils-parsexs/blob/29e5a1ef39513ff5815b3faf8fe8b47d68c72e9a/lib/ExtUtils/ParseXS.pm#L175


But passed typemaps are not actually taking precedence unless it's named 'typemap' and appears in 
one of the locations listed here:

https://github.com/Perl-Toolchain-Gang/extutils-parsexs/blob/29e5a1ef39513ff5815b3faf8fe8b47d68c72e9a/lib/ExtUtils/ParseXS/Utilities.pm#L102


The problem is, standard_typemap_locations( \@INC ), shown here

https://github.com/Perl-Toolchain-Gang/extutils-parsexs/blob/29e5a1ef39513ff5815b3faf8fe8b47d68c72e9a/lib/ExtUtils/ParseXS/Utilities.pm#L275

is push()ed to the end of the typemap array, @tmap, in process_typemaps() instead of being unshift()ed, which would allow supplied typemaps 
to properly take precedence.

The reason this problably went unnoticed for so long is most likely due to the problem only occuring when overriding a typemap defined 
in `perl -V:privlib`/ExtUtils/typemap.

For example, when './typemap-foo' and `perl -V:privlib`/ExtUtils/typemap both define a mapping for the same data type, the definition
in `perl -V:privlib`/ExtUtils/typemap, which is processed after './typemap-foo', is the one that is used because the last one wins.

When './typemap-foo' defines type mappings that are not found in `perl -V:privlib`/ExtUtils/typemap (or anywhere else) then they are 
merged in just fine since there aren't any existing mappings `perl -V:privlib`/ExtUtils/typemap for that entry to clobber it.


Below are some simple one-liner examples that demonstrate the problem:


### ExtUtils::ParseXS::Utilities::process_typemaps([@ARGV], pwd) in the examples below are based on this line:
### https://github.com/Perl-Toolchain-Gang/extutils-parsexs/blob/29e5a1ef39513ff5815b3faf8fe8b47d68c72e9a/lib/ExtUtils/ParseXS.pm#L175
### Run this in an empty directory:

# Linux/Unix

$ \rm typemap*; echo -e 'TYPEMAP\nwchar_t * T_WCHAR' > typemap; perl -MExtUtils::ParseXS::Utilities -E 'say ExtUtils::ParseXS::Utilities::process_typemaps([@ARGV], pwd)->as_string' typemap | grep -i wchar
wchar_t T_IV
wchar_t *       T_WCHAR

$ \rm typemap*; echo -e 'TYPEMAP\nwchar_t * T_WCHAR' > typemap-other; perl -MExtUtils::ParseXS::Utilities -E 'say ExtUtils::ParseXS::Utilities::process_typemaps([@ARGV], pwd)->as_string' typemap-other | grep -i wchar
wchar_t *       T_PV
wchar_t T_IV


# Win32:

>del typemap* & echo TYPEMAP > typemap & echo wchar_t * T_WCHAR >> typemap & perl -MExtUtils::ParseXS::Utilities -E "say ExtUtils::ParseXS::Utilities::process_typemaps([@ARGV], pwd)->as_string" typemap | grep -i wchar
wchar_t T_IV
wchar_t *       T_WCHAR

>del typemap* & echo TYPEMAP > typemap-other & echo wchar_t * T_WCHAR >> typemap-other & perl -MExtUtils::ParseXS::Utilities -E "say ExtUtils::ParseXS::Utilities::process_typemaps([@ARGV], pwd)->as_string" typemap-other | grep -i wchar
wchar_t *       T_PV
wchar_t T_IV

Leont commented:
> The reason this probably went unnoticed for so long is most likely due to the problem only occurring when overriding a typemap defined in perl -V:privlib/ExtUtils/typemap.

This sounds like an [old bug] that I thought had been fixed.

bay-max1 committed:
Check if user-supplied typemaps take precendence

Two new tests added to make sure that the entries in last-most typemap files listed in the array-ref passed to process_typemaps() actually take precedence over the entries in the files that appear before.

mohawk2 commented:
Since I gather this has a test that fails without the patch, I recommend this be merged & released. What do others think?

bay-max1 committed:
Use cmp_ok() instead of like() for clearer output.
